src/nbl/ext/FFT/FFT.cpp

```diff
@@ -0,0 +1,368 @@
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_DESCRIPTOR_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_DESCRIPTOR_DEFINED_
+layout(set=_NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_, binding=_NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_) restrict buffer OutputBuffer
+{
+	nbl_glsl_ext_FFT_output_t outData[];
+};
+#endif
+
+// Get/Set Data Function
+
+vec2 nbl_glsl_ext_FFT_getData(in uvec3 coordinate, in uint channel)
+{
+	vec2 retValue = vec2(0, 0);
+#if USE_SSBO_FOR_INPUT > 0
+	uvec3 dimension = pc.dimension;
+	uint index = channel * (dimension.x * dimension.y * dimension.z) + coordinate.z * (dimension.x * dimension.y) + coordinate.y * (dimension.x) + coordinate.x;
+	retValue = inData[index].complex_value;
+#else
+	vec4 texelValue= texelFetch(inputImage, ivec2(coordinate.xy), 0);
+	retValue = vec2(texelValue[channel], 0.0f);
+#endif
+	return retValue;
+}
+
+void nbl_glsl_ext_FFT_setData(in uvec3 coordinate, in uint channel, in vec2 complex_value)
+{
+	uvec3 dimension = pc.padded_dimension;
+	uint index = channel * (dimension.x * dimension.y * dimension.z) + coordinate.z * (dimension.x * dimension.y) + coordinate.y * (dimension.x) + coordinate.x;
+	outData[index].complex_value = complex_value;
+}
+
+void main()
+{
+	nbl_glsl_ext_FFT();
+}
+
+)===";
+
+	const size_t extraSize = 32 + 32 + 32 + 32;
+
+	const uint32_t maxItemsPerThread = core::ceil(float(maxPaddedDimensionSize) / DEFAULT_WORK_GROUP_X_DIM);
+	const uint32_t useSSBO = (DataType::SSBO == inputType) ? 1 : 0;
+	auto shader = core::make_smart_refctd_ptr<ICPUBuffer>(strlen(sourceFmt)+extraSize+1u);
+	snprintf(
+		reinterpret_cast<char*>(shader->getPointer()),shader->getSize(), sourceFmt,
+		useSSBO,
+		DEFAULT_WORK_GROUP_X_DIM,
+		maxPaddedDimensionSize,
+		maxItemsPerThread
+	);
+
+	auto cpuSpecializedShader = core::make_smart_refctd_ptr<ICPUSpecializedShader>(
+		core::make_smart_refctd_ptr<ICPUShader>(std::move(shader),ICPUShader::buffer_contains_glsl),
+		ISpecializedShader::SInfo{nullptr, nullptr, "main", asset::ISpecializedShader::ESS_COMPUTE}
+	);
+	
+	auto gpuShader = driver->createGPUShader(nbl::core::smart_refctd_ptr<const ICPUShader>(cpuSpecializedShader->getUnspecialized()));
+	
+	auto gpuSpecializedShader = driver->createGPUSpecializedShader(gpuShader.get(), cpuSpecializedShader->getSpecializationInfo());
+
+	return gpuSpecializedShader;
+}
+
+void FFT::defaultBarrier()
+{
+	COpenGLExtensionHandler::pGlMemoryBarrier(GL_SHADER_STORAGE_BARRIER_BIT);
+}
+
+// Kernel Normalization
+
+core::smart_refctd_ptr<video::IGPUSpecializedShader> FFT::createKernelNormalizationShader(video::IVideoDriver* driver)
+{
+	const char* sourceFmt =
+R"===(#version 430 core
+
+layout(local_size_x=%u, local_size_y=1, local_size_z=1) in;
+ 
+struct nbl_glsl_ext_FFT_output_t
+{
+	vec2 complex_value;
+};
+
+layout(set=0, binding=0) restrict readonly buffer InBuffer
+{
+	nbl_glsl_ext_FFT_output_t in_data[];
+};
+
+layout(set=0, binding=1) restrict buffer OutBuffer
+{
+	nbl_glsl_ext_FFT_output_t out_data[];
+};
```

Comment:
a small note... you could do this in-place (normalize a buffer without having a copy) iby using the driver's GPU to GPU copy ability to copy the 8 bytes of `in_data[0]` to a temporary buffer used as an UBO in the shader and issue a barrier between the small copy and dispatch

Then you'd only have one descriptor in the set and no double copies of the buffer

Reply:
Could save a lot of memory in lets say the blooming of 8k HDR images (512mb per padded frame) because the kernel would need to be padded to next PoT size of 8k too